src/ui/Windows/MainView.cpp

```diff
@@ -63,6 +63,12 @@ void DboxMain::DatabaseModified(bool bChanged)
 {
   PWS_LOGIT_ARGS("bChanged=%s", bChanged ? L"true" : L"false");
 
+  // If called from worker thread, invoke on GUI thread.
+  if (!IsGuiThread()) {
```

Comment:
I could be way off base here (not a UI or MFC expert), but is it possible to prevent this situation from happening in the first place? Seems like where this method is called from isn't random - probably some other code is invoking this on the wrong thread.

Reply:
> I could be way off base here (not a UI or MFC expert), but is it possible to prevent this situation from happening in the first place? Seems like where this method is called from isn't random - probably some other code is invoking this on the wrong thread.

PWSCore's apparent design is to call registered observers from its worker thread. DboxMain must therefore expect it will certainly be called from a non-UI non-MFC worker thread. It's up to DboxMain to ensure it updates the UI correctly. The typical way to do this is dispatching work to the UI thread which is what this change does.

Whether some sort of larger effort could rethink the overall design, the need for UI updates here, is always a potential. 

My goal with this PR was to remove noise interfering with debugging new feature work, and to follow any apparent contracts where reasonable. In this case, MFC asks not to have its objects used from non-MFC threads unless allowed by design... this change is about complying given assertions indicating at least some such usage was not by design and potentially problematic.

Beyond the assertions, though, the rules still seem apparently broken if we do not dispatch to the UI thread wholly lest we spend a lot of time picking apart each line of each potential code path to see which MFC usage is fine cross-thread or not... hardly worthwhile... better is to just take hint from assertion, follow its rules, dispatch to UI, avoid asserts, and then use saved brain cycles on more useful feature work, if that makes sense.

Possibly helpful...

[Accessing MFC Objects from Non-MFC Threads](https://learn.microsoft.com/en-us/cpp/parallel/multithreading-programming-tips?view=msvc-170#_core_accessing_mfc_objects_from_non.2d.mfc_threads)

[Windows Handle Maps](https://learn.microsoft.com/en-us/cpp/parallel/multithreading-programming-tips?view=msvc-170#_core_windows_handle_maps)